### PR TITLE
Export traces to Jaeger and update OpenTelemetry

### LINKED ISSUE #1113: Export traces in OpenTelemetry format and then to Azure Monitor
/kind feature
/help

**Describe the solution you'd like**
As a user of Cluster API Provider Azure, I would like to be able to export trace spans in OpenTelemetry format so that they can then be exported to Azure Monitor via the [Azure Monitor Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/exporter/azuremonitorexporter).

Currently, when a developer runs `make tilt-up`, Jaeger is started on http://localhost:8080 and a Jaeger "all-in-one" sidecar is injected to into the controller pod. The CAPZ controller sends traces to the Jaeger sidecar for direct analysis.

The desired solution is to be able to run `make tilt-up` and have https://github.com/open-telemetry/opentelemetry-collector or https://github.com/open-telemetry/opentelemetry-operator setup with the CAPZ controller exporting traces to a collector configured to export to Azure Monitor using the Azure Monitor OpenTelemetry Exporter.

We should also document the components and configuration needed to create a CAPZ cluster with these components installed for non-development scenarios. (@mboersma, this might be an interesting idea for a CLI gesture)

**Anything else you would like to add:**
The first tracing PR #1019 added Jaeger support. It is a good starting point to learn from, but lacks the flexibility of design which comes with using the OpenTelemetry collector. Separating the collector from the exporter / analysis will enable users of CAPZ to have more control over their stack of tools.

### PR DESCRIPTION
**Pull request by mboersma**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

See #1439 for part one of these changes.

Adds (back) a Jaeger all-in-one Pod for viewing CAPZ controller traces locally (in addition to App Insights). Updates all opentelemetry-go code to their latest split releases, v1.0.0-RC2 and v0.22.0. Adds "View traces" and "View metrics" links on the appropriate Tilt resources, as well as an "App Insights" link if `AZURE_INSTRUMENTATION_KEY` was set. Adds and updates lots of observability docs.

**Which issue(s) this PR fixes**:

Fixes #1113
Depends on #1439 (or closes it, if we merge this directly)

**Special notes for your reviewer**:

#1439 is already `XXL` and so is this additional changeset. I think we should merge #1439 first and then I can rebase this for reviewing separately. Just wanted to get it out there for any feedback.

I thought it was important to have a local (Jaeger) solution for viewing traces, rather than require App Insights. Please let me know if the Tilt warning approach seems too heavy-handed.

Are there unit tests that could be written here (that would actually be worthwhile)?

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests
- [ ] add a readme.md for hack/observability/jaeger

**Release note**:

```release-note
Export traces to Jaeger and update OpenTelemetry
```

### CODE REVIEW

File: version/version.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func Get() Info {
 		GoVersion:       runtime.Version(),
 		Compiler:        runtime.Compiler,
 		Platform:        fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
-		AzureSdkVersion: fmt.Sprintf("v%s", azuresdkversion.Number),
```
Comment (mboersma) [inline version/version.go:61]: This was just a bug I noticed--`azuresdkversion.Number` is already "v55.2.0", so no there's no need to prepend a "v."

File: pkg/ot/traces.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	return nil
+}
+
+// otlpTracerProvider initializes an OTLP exporter and configures the corresponding tracer provider.
+func otlpTracerProvider(ctx context.Context, url string) (*sdktrace.TracerProvider, error) {
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			semconv.ServiceNameKey.String("capz"),
+			attribute.String("exporter", "otlp"),
+			attribute.String("version", version.Get().String()),
```
Comment (mboersma) [inline pkg/ot/traces.go:65]: `version.Get().String()` returns "" for me in `make tilt-up`. You can see the same in the controller logs. Is this expected?

Also, I just added "version" and "azuresdk.version" here as an example of trace attributes, and to see how they show up in App Insights and Jaeger, but we may not want to add these.

**Comment by mboersma** [comment]
I retested this PR and it's working well. I think three things remain:

- [ ] add a screenshot of Jaeger traces
- [ ] check that there are resource requests/limits on all observability components
- [ ] are there any opportunities here for unit test coverage?

**Comment by mboersma** [comment]
I rebased this, updated to otel v1.0.0-RC2/v0.22.0, and retested locally.  It's working well, so it's ready for review IMHO.

**Comment by mboersma** [comment]
/retest

A test cluster failed to provision, unrelated to this change AFAICT.

File: pkg/ot/traces.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		otlptracegrpc.WithInsecure(),
+		otlptracegrpc.WithEndpoint(url),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create otlp trace exporter")
+	}
+
+	bsp := sdktrace.NewBatchSpanProcessor(traceExporter)
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
```
Comment (devigned) [inline pkg/ot/traces.go:83]: We are going to want to think about this in the future. Perhaps, we should have some configuration around this to allow for a percentage based sampling. I don't think we should gate this PR on this, but just wanted to bring it up.

**Review by devigned — APPROVED** [review]
/lgtm

I did leave one piece of feedback about always sampling and that folks in production will likely want some control over the amount of sampling they are doing, rather than all or nothing as configuration allows today. This is something we should add in the future, but shouldn't gate this PR.

Awesome work on the readme(s)!

File: pkg/ot/traces.go (hunk 3 above)
Comment (mboersma) [reply pkg/ot/traces.go:83]: 👍🏻 Totally agree, but since this initial implementation is just for devs, I think it's ok to hard-code some things for now.

**Comment by mboersma** [comment]
I've had many e2e failures on this PR, but they all seem to be of the "timed out waiting for cluster to provision" variety which doesn't seem attributable to this PR. Are we just having a flaky time currently in CI or is there something I should dig into?

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/approve

🙌  +1, great work on the readmes
Also, nice touch with the tilt links

**Pull request merged**